### PR TITLE
Split the current Array into Array and Mutable.Array

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

**Comment by adithyaov** [comment]
* `default(0)` is this PR `array-immu3`
* `default(1)` is master

There seems to be no major change. A little improvement though.

```
Generating reports for array...
Array(time)
Benchmark                                     default(0)(μs) default(1) - default(0)(%)
--------------------------------------------- -------------- --------------------------
array/elimination/length . IsList.toList              195.31                     +18.91
array/elimination/foldl'                              150.02                     +18.26
array/generation/writeN . IsString.fromString         899.31                     +14.65
array/generation/writeN . intFromTo                   163.41                     +14.51
array/generation/writeN . IsList.fromList            1211.24                     +11.25
array/transformationX4/map                            629.29                     +10.02
array/generation/fromList . intFromTo                1247.38                      +9.38
array/transformationX4/scanl1'                        622.54                      +8.94
array/elimination/read                                143.33                      +7.95
array/generation/show                                5136.29                      +4.74
array/generation/write . intFromTo                    215.57                      +1.55
array/generation/writeN . fromList                   1121.08                      +1.44
array/elimination/id                                  148.59                      +0.98
array/generation/read                              240136.42                      +0.95
array/transformation/scanl'                           293.61                      -1.24
array/elimination//=                                  139.14                      -1.35
array/elimination/==                                  140.94                      -2.92
array/transformation/map                              288.45                      -3.30
array/transformationX4/scanl'                         691.15                      -4.96
array/transformation/scanl1'                          325.21                      -7.19
array/elimination/min                                 303.00                      -8.33
array/elimination/<                                   215.27                     -11.73
array/generation/writeN . unfoldr                     185.81                     -16.06
array/elimination/toStreamRev                         177.06                     -22.84

```

File: src/Streamly/Internal/Memory/Mutable/Array/Types.hs [hunk 1]
```diff
@@ -0,0 +1,1366 @@
...
+#ifdef DEVBUILD
+    Storable a =>
+#endif
+    Array
+    { aStart :: {-# UNPACK #-} !(ForeignPtr a) -- first address
+    , aEnd   :: {-# UNPACK #-} !(Ptr a)        -- first unused address
+    , aBound :: {-# UNPACK #-} !(Ptr a)        -- first address beyond allocated memory
+    }
+
+mutableArray :: ForeignPtr a -> Ptr a -> Ptr a -> Array a
```
Comment (adithyaov) [inline src/Streamly/Internal/Memory/Mutable/Array/Types.hs]: Add a `INLINE` here

**Comment by harendra-kumar** [comment]
Please remove the old commented out code.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 2]
```diff
@@ -4348,19 +4349,19 @@ tapAsync f (Stream step1 state1) = Stream step TapInit
...
+            rh1 <- liftIO $ RB.unsafeInsert rb rh a
+            return $ Tuple3' rb rh1 (i + 1)
+        initial = fmap (\(a, b) -> Tuple3' a b (0 :: Int)) $ liftIO $ RB.new n
+        done (Tuple3' rb rh i) = do
+            arr <- liftIO $ MA.newArray n
+            foldFunc i rh snoc' arr rb
+        snoc' b a = liftIO $ MA.unsafeSnoc b a
+        foldFunc i
+            | i < n = RB.unsafeFoldRingM
+            | otherwise = RB.unsafeFoldRingFullM
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:4364]: change indentation before merge

**Comment by harendra-kumar** [comment]
@adithyaov can you squash the changes and cleanup commit messages. I will just make any review changes directly on top of it to commit it ASAP.

**Comment by adithyaov** [comment]
Done. Perhaps we can rename the mutable array according to the new module conventions.

File: src/Streamly/Internal/Memory/Array/Types.hs [hunk 3]
```diff
@@ -15,28 +15,33 @@ module Streamly.Internal.Memory.Array.Types
...
+    , unsafeFreezeWithShrink
+    , unsafeThaw
+
     -- * Construction
-    , withNewArray
-    , newArray
-    , unsafeSnoc
-    , snoc
-    , spliceWithDoubling
+--    , withNewArray
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Memory/Array/Types.hs:24]: Why are these commented? Why aren't they removed?

File: src/Streamly/Internal/Memory/Array/Types.hs [hunk 4]
```diff
@@ -15,28 +15,33 @@ module Streamly.Internal.Memory.Array.Types
...
+--    , fromString#
     , fromList
     , fromListN
     , fromStreamDN
     -- , fromStreamD
 
     -- * Streams of arrays
     , fromStreamDArraysOf
-    , FlattenState (..) -- for inspection testing
+    , MA.FlattenState (..) -- for inspection testing
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Memory/Array/Types.hs:40]: Why are we exporting Mutable array stuff from immutable array?

File: src/Streamly/Internal/Memory/Array/Types.hs (hunk 3 above)
Comment (adithyaov) [reply src/Streamly/Internal/Memory/Array/Types.hs:24]: No specific reason. I got into a habit of commenting code ove removing. In this case they should be removed.

File: src/Streamly/Internal/Memory/Array/Types.hs (hunk 4 above)
Comment (adithyaov) [reply src/Streamly/Internal/Memory/Array/Types.hs:40]: Yeah, I should remove these.

**Pull request merged**